### PR TITLE
Use NamedArgumentConstructorAnnotation for annotations

### PR DESCRIPTION
**Pull request by malarzm**
Brings https://github.com/doctrine/annotations/pull/369 to the ODM, here's what can be done later: https://github.com/doctrine/orm/pull/8266

|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | kind of but not really (see UPGRADE notes)
| Fixed issues | n/a

### CODE REVIEW

File: UPGRADE-2.2.md [hunk 1]
```diff
@@ -9,9 +9,20 @@ to the `Query` class.
 The `Doctrine\ODM\MongoDB\Aggregation\Builder::execute()` method was deprecated
 and will be removed in ODM 3.0.
 
-## Document indexes (annotations)
+## Aannotations
+
+### Base class
+
+All ODM's annotations no longer extend the `Doctrine\Common\Annotations\Annotation`
+and may require passing arguments to their constructors.
```
Comment (malarzm) [inline UPGRADE-2.2.md:17]: missing `class`

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/AbstractField.php [hunk 2]
```diff
@@ -4,28 +4,44 @@
...
 
-abstract class AbstractField extends Annotation
+abstract class AbstractField implements NamedArgumentConstructorAnnotation
 {
     /** @var string */
     public $name;
 
-    /** @var string */
-    public $type = 'string';
+    /** @var string|null */
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/AbstractField.php:15]: `|null` comes from `@Id`

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/AlsoLoad.php [hunk 3]
```diff
@@ -4,15 +4,24 @@
...
  * @Annotation
  */
-final class AlsoLoad extends Annotation
+final class AlsoLoad implements NamedArgumentConstructorAnnotation
 {
-    /** @var string */
+    /** @var string|string[] */
+    public $value;
+
+    /** @var string|null */
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/AlsoLoad.php:19]: I tried to fix all the wrong docblocks by the way

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/ChangeTrackingPolicy.php [hunk 4]
```diff
@@ -4,13 +4,20 @@
...
  */
-final class ChangeTrackingPolicy extends Annotation
+final class ChangeTrackingPolicy implements NamedArgumentConstructorAnnotation
 {
+    /** @var string */
+    public $value;
+
+    public function __construct(string $value)
+    {
+        $this->value = $value;
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/ChangeTrackingPolicy.php:21]: Idea for later PR (or commit): add validation to ctors

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Document.php [hunk 5]
```diff
@@ -4,17 +4,19 @@
...
  * @Annotation
  */
-final class Document extends AbstractDocument
+final class Document extends AbstractDocument implements NamedArgumentConstructorAnnotation
 {
     /** @var string|null */
     public $db;
 
-    /** @var string|null */
+    /** @var array|string|null */
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Document.php:19]: `array` is used for capped collection's config

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/File/ChunkSize.php [hunk 6]
```diff
@@ -11,12 +11,14 @@
...
-    public $name = 'chunkSize';
-
-    /** @var string */
-    public $type = 'int';
-
-    /** @var bool */
-    public $notSaved = true;
+    public function __construct(
+        string $name = 'chunkSize',
+        string $type = 'int',
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/File/ChunkSize.php:16]: Does it even make sense to allow overriding anything but `$name` here?

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/ReferenceMany.php [hunk 7]
```diff
@@ -35,7 +32,7 @@ final class ReferenceMany extends AbstractField
     /** @var string|null */
     public $defaultDiscriminatorValue;
 
-    /** @var string[]|null */
+    /** @var string|string[]|null */
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/ReferenceMany.php:35]: turns out using string is also fine, it's converted to an array later while mapping field

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/File/ChunkSize.php (hunk 6 above)
Comment (alcaeus) [reply lib/Doctrine/ODM/MongoDB/Mapping/Annotations/File/ChunkSize.php:16]: No: only the name can be configured, the rest is specified in the [GridFS Specification](https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst).

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Inheritance.php [hunk 8]
```diff
@@ -1,27 +0,0 @@
...
-namespace Doctrine\ODM\MongoDB\Mapping\Annotations;
-
-use Doctrine\Common\Annotations\Annotation;
-
-/**
- * Specifies inheritance mapping for a document
- *
- * @Annotation
- */
-final class Inheritance extends Annotation
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Inheritance.php:14]: Given that this class is `final` and never checked for, it should be safe to remove it. I'm sure there's someone out there using it, not realising that it didn't do anything.

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/MappedSuperclass.php [hunk 9]
```diff
@@ -4,14 +4,21 @@
...
 /**
  * Specifies a parent class that other documents may extend to inherit mapping
  * information
  *
  * @Annotation
  */
-final class MappedSuperclass extends AbstractDocument
+final class MappedSuperclass extends AbstractDocument implements NamedArgumentConstructorAnnotation
 {
     /** @var string */
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/MappedSuperclass.php:17]:
```suggestion
    /** @var string|null */
```

**Review by alcaeus — COMMENTED** [review]
As a general theme, I'd advocate to rename the `$value` arguments in constructors to something more meaningful. For example, in the `AlsoLoad` annotation, `$value` doesn't really tell me what this should be, especially with `$name` being the other field. This is especially true with these names being somewhat relevant to BC when people use named arguments.

**Comment by malarzm** [comment]
Getting back to the PR got me so long somebody opened https://github.com/doctrine/annotations/issues/396 which actually is great :)

**Comment by IonBazan** [comment]
@malarzm Since #2344 has been merged and contains similar changes, should we close this one or would you like to backport the `@NamedArgumentConstructorAnnotation` changes to `2.2.x`?

**Comment by malarzm** [comment]
Nah, let's close this one 👍

**Pull request closed without merging**